### ISSUE TITLE
Half precision display not exactly correct

### ISSUE THREAD

**Issue by ZisIsNotZis** [numpy/numpy#12613]
I found this problem some days ago. I understand this is not a serious problem, but I feel it is worth mentioning here?

### Reproducing code example:
    
    import numpy as np
    n = np.uint16(0b0111101111111111).view('e')
    print(n, n.astype('f'))

### Output:

    65500.0 65504.0

### Error message:

There are no error message, but `float16` `0b0111101111111111` should be around `65504.0` instead of `65500.0`.

### Numpy/Python version information:

    1.15.4 3.6.6 |Anaconda, Inc.| (default, Oct  9 2018, 12:34:16) 
    [GCC 7.3.0]

**Comment by eric-wieser** [comment]
cc @ahaldane. I think what's relevant here is the difference between "the exact decimal this float16 represents" and "the decimal with the minimum number of digits needed to uniquely identify the float16".

**Comment by ahaldane** [comment]
Right. This is the intended behavior currently.

That is, currently the default float-printing strategy is to only print the leading digits necessary to represent the value unambiguously, and then pad with zeros. This way only the "significant" digits are shown. It turns out that '65500' is the "shortest" way or representing "65504" since the 4 is unnecessary:
```python
>>> np.float16('65500') == np.float16('65504')
```
You can see the infinite-precision value by doing:
```python
>>> np.format_float_positional(n, precision=20, unique=False)
'65504.00000000000000000000'
```

I admit in your case looks a little weird, and maybe ideally when the sig-figs are to the left of the decimal point we should print the exact value up to the decimal point. On the other hand, when the sigfigs are to the right of the decimal points it makes sense to drop unneeded digits. 

Here are some more examples:
```python
>>> n = np.float16('50000')
>>> repr(n), np.format_float_positional(n, precision=20, unique=False)
('50000.0', '49984.00000000000000000000')
>>> n = np.float16('5e-4')
>>> repr(n), np.format_float_positional(n, precision=20, unique=False)
('0.0005', '0.00050020217895507812')
```

**Comment by eric-wieser** [comment]
It would be nice to have a float-printing mode that shows all non-zero digits, which.

I think given the extremely narrow range of float16, it might make sense to use a different default printing behavior to other floats.

**Comment by ahaldane** [comment]
True, I don't think we've ever put that much thought into what the default should be.

The eval example happens for other non-float64 floats though. Eg float128 won't satisfy `eval(repr(float128_val)) != float128_val` in most cases because the python float will lose digits.

Another possibility is to change the exponential-mode cutoff for different float types.  Currently all float scalars switch at 1e16, which is near the point where float64 can no longer represent integer values. Maybe we should switch float16 around 1e3 and float32 around 1e7, which is near where they no longer represent integers. 

There re back-compat issues to worry about too for such changes.

**Comment by ZisIsNotZis** [comment]
Oh, indeed, the next largest number a `float16` can represent is `65472`. That's why `4` in `65504` is not necessary. Anyway, I found this when I was trying to convert `float16` to `float10` and this kind of confused me for a while when debugging (along with floating point are also little endian).

I guess it is not that necessary to fix then? Or simply use fully-display strategy when outputting a single number?

**Comment by ahaldane** [comment]
@ZisIsNotZis well it's useful to bring it up anyway, since I'm not sure we thought about it much before. We have new printing code as of numpy 1.14 or 1.15 which made floats print with better sigfigs, but didnt consider the left side of the decimal point thoroughly.

Actually, after sitting on it I think it would be a good idea to change the exponential-mode (scientific notation) cutoff for different float types as I said in my last comment. Then the value `np.float16('65504')` would print as `6.55e+04` which is much better than `'65500.0` since it is more clearly rounded, and rather than printing "incorrect" digits it merely omits them.

For float64, the current cutoff of 1e16 is good because we never end up printing "wrong" digits to the left of the decimal point, in the way you found float16 does. That because all values up to `2**53` (`10**15.954`) are exactly printable in non-scientific notation due to 53-bit mantissa, and the values from `10**15.954` to `10**16` don't have any extra digits in base-10 so no zero-padding occurs. That means we never print 0s in places that are really not zero.

For float16 with 11-bit mantissa, I think we should start scientific notation at `10**4`, since all values up to `2**11` (`10**3.311`) are exactly representable and the values from `10**3.311` to `10**4` don't need any extra digits in base-10.

**Comment by eric-wieser** [comment]
Seems like a convincing argument to me, @ahaldane. Want to make a patch?

**Comment by eric-wieser** [comment]
One thing that worries me is that the behavior will still seem wrong if users override the scientific cutoff, even if we choose a more sensible default
